problemtools/verifyproblem.py

```diff
@@ -1221,35 +1232,60 @@ def __init__(self, problem):
     def __str__(self):
         return 'submissions'
 
-    def check_submission(self, sub, args, expected_verdict, timelim_low, timelim_high):
+    def check_submission(self, sub, args, expected_verdict, timelim, timelim_low, timelim_high):
         desc = '%s submission %s' % (expected_verdict, sub)
-        (result1, result2) = self._problem.testdata.run_submission(sub, args, timelim_low, timelim_high)
-
-        if result1.verdict != result2.verdict:
-            r1, r2 = result1.verdict, result2.verdict
-            self.warning('%s sensitive to time limit: limit of %s secs -> %s, limit of %s secs -> %s' % (desc, timelim_low, r1, timelim_high, r2))
-
-        if result1.verdict == expected_verdict:
+        partial = False
+        if expected_verdict == 'PAC':
+            # For partially accepted solutions, use the low timelim instead of the real one,
+            # to make sure we have margin in both directions.
+            expected_verdict = 'AC'
+            partial = True
+            timelim = timelim_low
+
+        subinfo = {'desc': desc, 'sample_failures': []}
+        (result1, result2) = self._problem.testdata.run_submission(sub, args, timelim, timelim_high, subinfo)
+
+        if result1.verdict == 'AC' and expected_verdict == 'AC' and not partial and subinfo['sample_failures']:
+            res = subinfo['sample_failures'][0]
+            self.warning('%s got %s on sample: %s' % (desc, res.verdict, res))
+
+        if result1.verdict != result2.verdict or result1.score != result2.score:
+            r1, r2 = result1, result2 if result1.verdict == result2.verdict else result1.verdict, result2.verdict
+            self.warning('%s sensitive to time limit: limit of %s secs -> %s, limit of %s secs -> %s' % (desc, timelim, r1, timelim_high, r2))
+
+        if result1.verdict == expected_verdict and not (partial and self.fully_accepted(result1)):
             self.msg('   %s OK: %s' % (desc, result1))
-        elif result2.verdict == expected_verdict:
+            if expected_verdict == 'AC' and not partial and not self.fully_accepted(result1):
+                # For some heuristic problems, this is expected. Thus, only warn.
+                self.warning('%s did not attain full score (consider moving it to partially_accepted)' % desc)
+        elif result2.verdict == expected_verdict and not (partial and self.fully_accepted(result2)):
             self.msg('   %s OK with extra time: %s' % (desc, result2))
         else:
             self.error('%s got %s' % (desc, result1))
 
         return result1
 
+    def fully_accepted(self, result):
+        min_score, max_score = self._problem.testdata.get_score_range()
+        best_score = min_score if self._problem.config.get('grading')['objective'] == 'min' else max_score
+        return result.verdict == 'AC' and (not self._problem.is_scoring or result.score == best_score)
```

Comment:
I feel that this warning is going to be annoying for open-ended heuristic problems without a clear upper bound on the optimum.  So I suggest not giving the warning if max_score/min_score (whichever is applicable) is inf/-inf.

Reply:
Good point, will do